### PR TITLE
OpInfo: split, split_with_sizes

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #54261

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -6373,15 +6426,6 @@ def method_tests():
...
         ('squeeze', (), (0,), 'scalar', (True,), [0]),
-        ('split', (S, S, S), (2,), '', (True,)),
-        ('split', (S, S, S), (S, 1), 'dim', (True,), [1]),
-        ('split', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), 'size_list',
-            (True, 'aten::split_with_sizes')),
-        ('split', (S, S, S), ([int(S / 2), S - int(S / 2) * 2, int(S / 2)], 2), 'size_list_dim',
-            (True, 'aten::split_with_sizes'), [1]),
-        ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), '', (True,)),
-        ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3), 0],), 'size_0', (True, )),
-        ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), 'dim', (True, ), [1]),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:6384]: I assume that this cases was supposed to add a dim as well to the argument. Have added one in `sample_inputs_split_with_sizes`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -5152,6 +5190,21 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+    OpInfo('split',
+           dtypes=all_types_and_complex_and(torch.bfloat16, torch.half, torch.bool),
+           sample_inputs_func=sample_inputs_split,
+           supports_out=False,
+           assert_autodiffed=True),
+    OpInfo('split',
+           variant_test_name='list_args',
+           dtypes=all_types_and_complex_and(torch.bfloat16, torch.half, torch.bool),
+           sample_inputs_func=partial(sample_inputs_split, list_args=True),
+           supports_out=False),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:5202]: Looking at these entries,
```
('split', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), 'size_list',
            (True, 'aten::split_with_sizes')),
 ('split', (S, S, S), ([int(S / 2), S - int(S / 2) * 2, int(S / 2)], 2), 'size_list_dim',
            (True, 'aten::split_with_sizes'), [1]),
```
and 
https://github.com/pytorch/pytorch/blob/cbd122780951cb23025a80ab5d43f81c9cc2077b/torch/testing/_internal/common_methods_invocations.py#L6210

tried adding to the entry with `list_args`
```python
assert_autodiffed=True,
autodiff_nonfusible_nodes=['aten::split_with_sizes']
```
but jit consistency test failed with,
```
Failure in testing nodes' autodifferentiation. One or more nodes were expected to be autodiffed, but were not found in specified fusible/nonfusible DifferentiableGraph groups. 
Specifically:
  ['aten::split_with_sizes'] were not in one of the DifferentiableGraphs when they were expected to be. Did you intend for these nodes to be autodiffed? If not, remove them from the list of nonfusible nodes.
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6384]: I think the test generator from method_tests understood the dim argument and would programmatically generate some options for it

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -3212,6 +3212,45 @@ def generator():
 
     return list(generator())
 
+
+def sample_inputs_split(op_info, device, dtype, requires_grad, list_args=False, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3216]: Style nit: make `list_args` kwarg-only

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5203]: Great attention to detail as always, @kshitij12345 

I'm not sure what's going on here. I made a sample program:

```
import torch
S = 5
t = torch.randn((S, S, S), requires_grad=True)

def foo(t):
  S = 5
  return t.split([int(S / 2), S - int(S / 2) * 2, int(S / 2)])

scripted_foo = torch.jit.script(foo)
scripted_foo.graph_for(t)

traced_foo = torch.jit.trace(foo, (t,))
traced_foo.graph
```

And I can see "split" becomes "split_with_sizes" in the traced graph, but not the scripted graph, and it doesn't look like it's part of a differentiable subgraph in either case.

Looking at the code it does seem like there's some special analysis that might convert split to split_with_sizes, but I can't see it being triggered, either:

https://github.com/pytorch/pytorch/blob/4cb534f92ef6f5b2ec99109b0329f93a859ae831/torch/csrc/jit/runtime/register_special_ops.cpp#L250

I wonder if the autodifferentiation part of method_tests() is even running? Would you submit a PR that just deletes the `(True, 'aten::split_with_sizes')` from split and see if anything fails? I tried to find the tests for it in the code and couldn't spot them.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6384]: I don't think that is the case.
For checking, I had added the following snippet after the linked line.
```python
if 'split_with' in test_name:
      print(test_name, new_args)
```
https://github.com/pytorch/pytorch/blob/82d714935ea7c54f9e5966f47adf4b8460eb2c3d/test/test_autograd.py#L5390

Output for the given sample was (not all args are same),
```
test_split_with_sizes_dim_neg0 ([1, 3, 1],)
test_split_with_sizes_dim ([1, 3, 1],)
test_split_with_sizes_dim_neg0_complex ([1, 3, 1],)
test_split_with_sizes_dim_complex ([1, 3, 1],)
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:5203]: Locally verified:
With the following diff
```patch
+        ('split', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), 'size_list',),
+        ('split', (S, S, S), ([int(S / 2), S - int(S / 2) * 2, int(S / 2)], 2), 'size_list_dim', (), [1]),
```

The tests pass locally.
```
python test/test_autograd.py -k _split
/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/backends/cudnn/__init__.py:73: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
................../home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/backends/cudnn/__init__.py:73: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
..................ssssssssssssssssss
----------------------------------------------------------------------
Ran 54 tests in 2.998s

OK (skipped=18)
```

Will submit a PR to verify the same on CI.
Thanks for digging into it and the suggestion :)

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6405]: Interesting; thanks for investigating. Good thing soon method_tests() will be gone and we won't have to wonder at its mysteries anymore

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

**Review by mruberry — APPROVED** [review]
Cool! Good investigation with this one.
